File: camera_sdk/Altair/samples/imageptrsimplest/imageptrsimplest.cpp
#include <stdio.h>
#include <stdlib.h>
#include "altaircam.h"

HAltaircam g_hcam = NULL;
unsigned g_total = 0;

static void __stdcall EventCallback(unsigned nEvent, void* pCallbackCtx)
{
    if (ALTAIRCAM_EVENT_IMAGE == nEvent)
    {
        AltaircamImagePtr ptr = { 0 };
        const HRESULT hr = Altaircam_PullImagePtr(g_hcam, 0, &ptr);
        if (FAILED(hr))
            printf("failed to pull imageptr, hr = 0x%08x\n", hr);
        else
        {
            /* After we get the image data, we can do anything for the data we want to do */
            printf("pull imageptr ok, total = %u, res = %u x %u, id = %u\n", ++g_total, ptr.ptrInfo->v3.width, ptr.ptrInfo->v3.height, ptr.id);
            Altaircam_PushImagePtr(g_hcam, ptr.id);
        }
    }
    else
    {
        printf("event callback: 0x%04x\n", nEvent);
    }
}

int main(int, char**)
{
    g_hcam = Altaircam_Open(NULL);
    if (NULL == g_hcam)
    {
        printf("no camera found or open failed\n");
        return -1;
    }

    HRESULT hr = Altaircam_StartPullModeWithCallback(g_hcam, EventCallback, NULL);
    if (FAILED(hr))
        printf("failed to start camera, hr = 0x%08x\n", hr);
    else
    {
        printf("press ENTER to exit\n");
        getc(stdin);
    }

    /* cleanup */
    Altaircam_Close(g_hcam);
    return 0;
}

File: camera_sdk/Altair/java/samples/imageptrsimplest/src/imageptrsimplest.java
import com.sun.jna.Structure;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.Arrays;

public class imageptrsimplest {
    static altaircam _cam = null;
    static int _total = 0;

    private static class ImplEventCallback implements altaircam.IEventCallback {
        /* the vast majority of callbacks come from altaircam.dll/so/dylib internal threads */
        @Override
        public void onEvent(int nEvent) {
            switch (nEvent) {
                case altaircam.EVENT_IMAGE:
                    try {
                        altaircam.ImagePtr ptrImage = new altaircam.ImagePtr();
                        _cam.PullImagePtr(0, ptrImage);
                        ++_total;
                        altaircam.FrameInfoV4 s = altaircam.Ptr2FrameInfoV4(ptrImage.ptrInfo);
                        System.out.printf("pull image ok: %d, width = %d, height = %d\n", _total, s.v3.width, s.v3.height);
                        _cam.PushImagePtr(ptrImage.id);
                    } catch (altaircam.HRESULTException e) {
                        System.out.println("pull image exception: " + e);
                    }
                    break;
                default:
                    System.out.println("event callback: " + nEvent);
                    break;
            }
        }
    }
    
    public static void main(String[] args) {
        altaircam.DeviceV2[] arr = altaircam.EnumV2();
        if (arr.length == 0)
            System.out.println("no camera found");
        else {
            System.out.println(arr[0].displayname + ": 0x" + Long.toHexString(arr[0].model.flag) + ", preview = " + arr[0].model.preview + ", still = " + arr[0].model.still);
            for (int i = 0; i < arr[0].model.res.length; ++i)
                System.out.println(arr[0].model.res[i].width + " x " + arr[0].model.res[i].height);

            _cam = altaircam.Open(arr[0].id);
            if (_cam != null) {
                try {
                    _cam.StartPullModeWithCallback(new ImplEventCallback());
                    System.out.println("Press Enter to exit");
                    try {
                        System.in.read();
                    } catch (IOException e) {
                    }
                } catch (altaircam.HRESULTException e) {
                    System.out.println("start camera exception: " + e);
                } finally {
                    _cam.close();
                    _cam = null;
                }
            }
        }
    }
}
